File: rust/tutorials/tour-of-restate-rust/src/part3/ticket_object.rs
use crate::auxiliary::ticket_status::TicketStatus;
use restate_sdk::prelude::*;

#[restate_sdk::object]
pub(crate) trait TicketObject {
    async fn reserve() -> Result<bool, HandlerError>;
    async fn unreserve() -> Result<(), HandlerError>;
    #[name = "markAsSold"]
    async fn mark_as_sold() -> Result<(), HandlerError>;
}

pub struct TicketObjectImpl;

impl TicketObject for TicketObjectImpl {
    // <start_reserve>
    async fn reserve(&self, ctx: ObjectContext<'_>) -> Result<bool, HandlerError> {
        let status: TicketStatus = ctx
            .get::<Json<TicketStatus>>("status")
            .await?
            .unwrap_or(Json(TicketStatus::Available))
            .into_inner();

        if let TicketStatus::Available = status {
            ctx.set("status", Json(TicketStatus::Reserved));
            Ok(true)
        } else {
            Ok(false)
        }
    }
    // <end_reserve>

    // <start_unreserve>
    async fn unreserve(&self, ctx: ObjectContext<'_>) -> Result<(), HandlerError> {
        let status: TicketStatus = ctx
            .get::<Json<TicketStatus>>("status")
            .await?
            .unwrap_or(Json(TicketStatus::Available))
            .into_inner();

        if let TicketStatus::Reserved = status {
            ctx.clear("status");
        }
        Ok(())
    }
    // <end_unreserve>

    // <start_mark_as_sold>
    async fn mark_as_sold(&self, ctx: ObjectContext<'_>) -> Result<(), HandlerError> {
        let status: TicketStatus = ctx
            .get::<Json<TicketStatus>>("status")
            .await?
            .unwrap_or(Json(TicketStatus::Available))
            .into_inner();

        if let TicketStatus::Reserved = status {
            ctx.set("status", Json(TicketStatus::Sold));
        }
        Ok(())
    }
    // <end_mark_as_sold>
}

File: go/tutorials/tour-of-restate-go/part3/ticketobject.go
package main

import (
	"github.com/restatedev/examples/go/tutorials/tour-of-restate-go/auxiliary"
	restate "github.com/restatedev/sdk-go"
)

type TicketObject struct{}

// <start_reserve>
func (TicketObject) Reserve(ctx restate.ObjectContext) (bool, error) {
	status, err := restate.Get[auxiliary.TicketStatus](ctx, "status")
	if err != nil {
		return false, err
	}

	if status == auxiliary.TicketStatusAvailable {
		restate.Set(ctx, "status", auxiliary.TicketStatusReserved)
		return true, nil
	} else {
		return false, nil
	}
}

// <end_reserve>

// <start_unreserve>
func (TicketObject) Unreserve(ctx restate.ObjectContext) error {
	status, err := restate.Get[auxiliary.TicketStatus](ctx, "status")
	if err != nil {
		return err
	}

	if status != auxiliary.TicketStatusSold {
		restate.Clear(ctx, "status")
	}

	return nil
}

// <end_unreserve>

// <start_mark_as_sold>
func (TicketObject) MarkAsSold(ctx restate.ObjectContext) error {
	status, err := restate.Get[auxiliary.TicketStatus](ctx, "status")
	if err != nil {
		return err
	}

	if status == auxiliary.TicketStatusReserved {
		restate.Set(ctx, "status", auxiliary.TicketStatusSold)
	}

	return nil
}

// <end_mark_as_sold>
